[Unified Text Replacement] Add support for undo-ing all replacements

rr-codes opened the pull request:
#### f82b26d2a0b7a540106b0bf407999ead4bba426c
<pre>
[Unified Text Replacement] Add support for undo-ing all replacements
<a href="https://bugs.webkit.org/show_bug.cgi?id=270138">https://bugs.webkit.org/show_bug.cgi?id=270138</a>
<a href="https://rdar.apple.com/123620291">rdar://123620291</a>

Reviewed by Aditya Keerthi and Megan Gardner.

* Source/WebKit/WebProcess/WebPage/UnifiedTextReplacementController.cpp:
(WebKit::UnifiedTextReplacementController::didEndTextReplacementSession):
(WebKit::UnifiedTextReplacementController::textReplacementSessionDidReceiveEditAction):
* Source/WebKit/WebProcess/WebPage/UnifiedTextReplacementController.h:

Canonical link: <a href="https://commits.webkit.org/275390@main">https://commits.webkit.org/275390@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/881b1093e5b5d9514c158fec173a9819eed0a3d2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/41674 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/20688 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/44055 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/44245 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/37767 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/23807 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/18018 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/34434 "Found 1 new test failure: media/track/media-element-enqueue-event-crash.html (failure)") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/42248 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/17603 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/35880 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/15101 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/15299 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/45627 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/37847 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/37220 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/40968 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/16489 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/13529 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/39386 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/18108 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/18164 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5581 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/17752 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas reviewed (commented):
LGTM

pxlcoder commented on Source/WebKit/WebProcess/WebPage/UnifiedTextReplacementController.cpp:
```diff
@@ -406,39 +407,50 @@ void UnifiedTextReplacementController::textReplacementSessionDidReceiveEditActio
-        if (m_replacements.isEmpty())
-            return;
+    case WebKit::WebTextReplacementData::EditAction::Undo: {
+        auto originalFragment = m_originalDocumentNodes.get(uuid);
```
I think we need to start using `Ref` rather than `auto` here, per current smart pointer guidelines. Similar instances throughout this change.

rr-codes replied:
```
In file included from /Users/richardrobinson/Build/Products/Debug/DerivedSources/WebKit/unified-sources/UnifiedSource118.cpp:2:
/Users/richardrobinson/safari/OpenSource/Source/WebKit/WebProcess/WebPage/UnifiedTextReplacementController.cpp:168:9: error: no viable constructor or deduction guide for deduction of template arguments of 'Ref'
  168 |     Ref liveRange = m_contextRanges.get(uuid);
In file included from /Users/richardrobinson/Build/Products/Debug/DerivedSources/WebKit/unified-sources/UnifiedSource118.cpp:1:
In file included from /Users/richardrobinson/safari/OpenSource/Source/WebKit/WebProcess/WebPage/PageBanner.cpp:27:
In file included from /Users/richardrobinson/safari/OpenSource/Source/WebKit/WebProcess/WebPage/PageBanner.h:28:
In file included from /Users/richardrobinson/safari/OpenSource/Source/WebKit/Shared/API/APIObject.h:29:
In file included from /Users/richardrobinson/Build/Products/Debug/usr/local/include/wtf/RefPtr.h:28:
/Users/richardrobinson/Build/Products/Debug/usr/local/include/wtf/Ref.h:85:5: note: candidate function [with T = WebCore::Range *, _PtrTraits = WTF::RawPtrTraits<WebCore::Range *>, RefDerefTraits = WTF::DefaultRefDerefTraits<WebCore::Range *>] not viable: expects an lvalue for 1st argument
   85 |     Ref(T& object)
/Users/richardrobinson/Build/Products/Debug/usr/local/include/wtf/Ref.h:123:5: note: candidate template ignored: couldn't infer template argument 'T'
  123 |     Ref(HashTableDeletedValueType) : m_ptr(PtrTraits::hashTableDeletedValue()) { }
/Users/richardrobinson/Build/Products/Debug/usr/local/include/wtf/Ref.h:126:5: note: candidate template ignored: couldn't infer template argument 'T'
  126 |     Ref(HashTableEmptyValueType) : m_ptr(hashTableEmptyValue()) { }
/Users/richardrobinson/Build/Products/Debug/usr/local/include/wtf/Ref.h:70:7: note: candidate template ignored: could not match 'Ref<T, type-parameter-0-1, type-parameter-0-2>' against 'MappedPeekType' (aka 'WebCore::Range *')
   70 | class Ref {
/Users/richardrobinson/Build/Products/Debug/usr/local/include/wtf/Ref.h:90:5: note: candidate template ignored: could not match 'Ref<T, type-parameter-0-1, type-parameter-0-2>' against 'MappedPeekType' (aka 'WebCore::Range *')
   90 |     Ref(const Ref& other)
/Users/richardrobinson/Build/Products/Debug/usr/local/include/wtf/Ref.h:95:38: note: candidate template ignored: could not match 'Ref<type-parameter-0-3, type-parameter-0-4>' against 'MappedPeekType' (aka 'WebCore::Range *')
   95 |     template<typename X, typename Y> Ref(const Ref<X, Y>& other)
]
```

getting these errors after doing so 😔

pxlcoder replied:
Try `RefPtr`?

megangardner reviewed (approved):
+1 to Aditya's comments

rr-codes committed:
[Unified Text Replacement] Add support for undo-ing all replacements
https://bugs.webkit.org/show_bug.cgi?id=270138
rdar://123620291

Reviewed by Aditya Keerthi and Megan Gardner.

* Source/WebKit/WebProcess/WebPage/UnifiedTextReplacementController.cpp:
(WebKit::UnifiedTextReplacementController::didEndTextReplacementSession):
(WebKit::UnifiedTextReplacementController::textReplacementSessionDidReceiveEditAction):
* Source/WebKit/WebProcess/WebPage/UnifiedTextReplacementController.h:

Canonical link: https://commits.webkit.org/275390@main
